server/s3_put_object_action.cc

```diff
@@ -841,6 +830,119 @@ void S3PutObjectAction::add_object_oid_to_probable_dead_oid_list_failed() {
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
+void S3PutObjectAction::_add_object_oid_to_probable_dead_oid_list(
+    std::shared_ptr<S3ObjectMetadata> old_object_metadata) {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  assert(!old_oid_str.empty());
+  if (number_of_parts != 0) {
+    // This object was uploaded previously in multipart fashion
+    add_part_object_to_probable_dead_oid_list(old_object_metadata,
+                                              probable_del_rec_list);
+    for (auto& probable_rec : probable_del_rec_list) {
+      probable_oid_list[probable_rec->get_key()] = probable_rec->to_json();
+    }
+  } else {
+    // Simple object
+    // prepending a char depending on the size of the object (size based
+    // bucketing of object)
+    S3CommonUtilities::size_based_bucketing_of_objects(
+        old_oid_str, old_object_metadata->get_content_length());
+
+    // key = oldoid + "-" + newoid
+    std::string old_oid_rec_key = old_oid_str + '-' + new_oid_str;
+    s3_log(S3_LOG_DEBUG, request_id,
+           "Adding old_probable_del_rec with key [%s]\n",
+           old_oid_rec_key.c_str());
+    old_probable_del_rec.reset(new S3ProbableDeleteRecord(
+        old_oid_rec_key, {0ULL, 0ULL}, old_object_metadata->get_object_name(),
+        old_object_oid, old_layout_id, old_object_metadata->get_pvid_str(),
+        bucket_metadata->get_object_list_index_layout().oid,
+        bucket_metadata->get_objects_version_list_index_layout().oid,
+        old_object_metadata->get_version_key_in_index(),
+        false /* force_delete */));
+    probable_oid_list[old_oid_rec_key] = old_probable_del_rec->to_json();
+    probable_del_rec_list.push_back(std::move(old_probable_del_rec));
+    old_obj_oids.push_back(old_object_oid);
+    old_obj_pvids.push_back(old_object_metadata->get_pvid());
+    old_obj_layout_ids.push_back(old_layout_id);
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutObjectAction::fetch_null_version_object_info() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  assert(!null_object_version_id.empty());
+  if (!null_object_metadata) {
+    null_object_metadata = object_metadata_factory->create_object_metadata_obj(
+        request, obj_list_idx_lo, obj_version_list_idx_lo);
+  }
+  null_object_metadata->set_version_id(null_object_version_id);
+  null_object_metadata->load(
```

Comment:
Which index table this metadata will be read from?

Reply:
Since I am setting the versionId, `load()` function will load the metadata from version_list_index.
Here is code snippet of `S3ObjectMetadata::load(...)` function for your reference - 
```
  if (object_version_id.empty()) {
    motr_kv_reader->get_keyval(
        object_list_index_layout, object_name,
        std::bind(&S3ObjectMetadata::load_successful, this),
        std::bind(&S3ObjectMetadata::load_failed, this));
  } else {
    motr_kv_reader->get_keyval(
        objects_version_list_index_layout, get_version_key_in_index(),
        std::bind(&S3ObjectMetadata::load_successful, this),
        std::bind(&S3ObjectMetadata::load_failed, this));
  }
```